tests/GetTests.hs

```diff
@@ -254,6 +263,102 @@ alterDistr' p1 p2 p3 =
+    parser = isolateLazy 100 $ do
+      skip 10
+      fail failStr
+      pure ()
+
+    failStr :: String
+    failStr = "no thanks"
+
+    go :: Result () -> IO ()
+    go r = case r of
+      Done () _ -> assertFailure "Impossible"
+      Fail failure _ -> unless (failStr `isInfixOf` failure) $ assertFailure "Wrong error!"
+      Partial cont -> assertFailure "Asked for more input!"
+
+instance Arbitrary LB.ByteString where
+  arbitrary = LB.fromChunks . pure . BS.pack <$> arbitrary
+
+newtype IsolationRes a = IRes (Either String a)
+  deriving Show
+
+-- Sometimes lazy and strict isolations return different errors,
+-- eg. when EOF is called before the end of an isolation which isn't prodided
+-- enough input.
+-- Strict sees it as a lack of bytes, Lazy sees it as a guard failure ("empty").
+instance Eq a => Eq (IsolationRes a) where
+  IRes a == IRes b = case (a, b) of
+    (Left e1, Left e2) -> e1 == e2 || errsEqAsymmetric e1 e2 || errsEqAsymmetric e2 e1
+    _ -> a == b
+    where
+      errsEqAsymmetric e1 e2 = "too few bytes" `isInfixOf` e1 && "empty" `isInfixOf` e2
+
+isolateAndIsolateLazy :: Int -> GetD -> LB.ByteString -> Property
+isolateAndIsolateLazy n parser' bs
+  = IRes (runGetLazy (isolate n parser) bs) === IRes (runGetLazy (isolateLazy n parser) bs)
+  where
+    parser = buildGet parser'
+
+isolateIsNotIncremental :: Assertion
+isolateIsNotIncremental = go (runGetPartial parser $ BS.replicate 11 0)
+  where
+    parser :: Get ()
+    parser = isolate 100 $ do
+      skip 10
+      fail failStr
+      pure ()
+
+    failStr :: String
+    failStr = "no thanks"
+
+    go :: Result () -> IO ()
+    go r = case r of
+      Done () _ -> assertFailure "Impossible"
+      Fail failure _ -> assertFailure $ "Strict isolate was incremental: " <> failure
+      Partial cont -> pure ()
+
+-- Checks return values, leftovers, fails for continuations
+assertResultsMatch :: Eq a => Result a -> (Maybe a, BS.ByteString) -> Assertion
+assertResultsMatch r1 r2 = case (r1, r2) of
+  (Partial _, _) -> assertFailure "Continuation received"
+  (Done a1 bs1, (Just a2, bs2)) -> do
+    unless (a1 == a2) $ assertFailure "Result mismatch"
+    unless (bs1 == bs2) $ assertFailure $ "Success leftover mismatch: " ++ show (bs1, bs2)
+  (Fail msg1 bs1, (Nothing, bs2)) ->
+    unless (bs1 == bs2) $ assertFailure $ "Failure leftovers mismatch: " ++ show (bs1, bs2)
+  _ -> assertFailure "Different result types"
+
+isolateLazyDeterminesLeftovers :: Assertion
+isolateLazyDeterminesLeftovers = do
+  assertResultsMatch (runGetPartial (isolateLazy 1 getWord8) "123") (Just $ toEnum $ fromEnum '1', "23")
+  assertResultsMatch (runGetPartial (isolateLazy 2 getWord8) "123") (Nothing, "3")
+  -- Note(414owen): I don't think this is the ideal behaviour, but it's the existing behaviour, so
+  -- I'll at least check that isolateLazy matches the behaviour of isolate...
+  assertResultsMatch (runGetPartial (isolate 2 $ fail "no thanks" *> pure ()) "123") (Nothing, "12")
```

Comment:
Can someone please explain why this is the bahaviour of `isolate`, rather than returning leftovers of `"3"`?

Reply:
To clarify: If we isolate a parser, I would expect it to either pass or fail, but either way to return leftovers of `drop n input`, where n is the isolation amount.

Reply:
`isolate` doesn't catch errors in the argument passed, so if there's a failure it will call the failure continuation with the current buffer. In this test, the buffer at that point is just the bytestring `"12"`. I think it would be reasonable to locally run the `m` argument to isolate so that it's possible to observe failures and put them in context. As your PR points out, it already strictly demands the bytes that it's isolating to, so it shouldn't ever hit any of the incremental machinery that exists.

Reply:
I'm just wondering if the user should be able to call another parser on the input stream, after a failure.

Currently, if we return the leftovers of the isolated parser at point of failure, there's no chance of continuing to use the input stream.

I guess the current isolation model is *prevent the parser from using beyond its designated bounds*, but that's it.

Reply:
This was somewhat intentional behavior in the original `isolate` function -- we made a lot of sacrifices in the name of performance, and not reconstructing the outer context in a failure is one of those. I think it would be reasonable to mirror the existing semantics, if only for consistency.

Reply:
Okay, great, that's what I've done, so I'll resolve this :)